File: mapping/src/Noise.java
public class Noise {
	// Starter function for the recursive function
	public static double getNoise(int channel, double resolution, double[] dimensions, double... coords) {
		return getNoise(channel, 1 / resolution, 0, dimensions.length, dimensions, coords);
	}
	
	// Recursively creates simple smooth noise for any number of dimensions
	private static double getNoise(int channel, double inverseResolution, double a, int numDimensions, double[] dimensions, double[] coords) {
		// The end condition that lerps between two random values offset by the channel
		if (--numDimensions == 0) {
			a = (a + coords[0]) * inverseResolution;
			
			long c = (long)a + randomLong(channel);
			
			return lerp(random(c), random(c + 1), a % 1);
		}
		
		// Assembles a unique coordinate (a) from the multidimensional coordinates iteratively
		double b = coords[numDimensions] * inverseResolution;
		
		double prod = 1;
		for (int i = 0; i < numDimensions; i++)
			prod *= dimensions[i] * inverseResolution;
		
		// lerps between the lower dimension noise values
		return lerp(getNoise(channel, inverseResolution, a + (int)b * prod, numDimensions, dimensions, coords),
					getNoise(channel, inverseResolution, a + ((int)b + 1) * prod, numDimensions, dimensions, coords),
					b % 1);
	}
	
	// Linear interpolation
	private static double lerp(double a, double b, double t) {
		double x = a + t * (b - a);
		return x * x * (3 - 2 * x);
	}
	
	// Random hash function
	public static long randomLong(long x) {
		x ^= 2747636419L;
		x *= 2654435769L;
		x ^= x >> 16;
		x *= 2654435769L;
		x ^= x >> 16;
		x *= 2654435769L;
		
		return x;
	}
	
	// Scales the random hash to a number between 0 and 1
	public static double random(long x) {
		return (double)(randomLong(x) >> 1) / Long.MAX_VALUE + 0.5;
	}
}

File: mapping/src/noise.js
// point is an iterable of floats describing the point in space at which to evaluate the noise function
// scale is an iterable describing the size of the total area in units
// dimensions is the number of dimensions (should always be the same as the length of point)
// evalAt is an internal value used for recursion (should always be 0)
// returns a value between 0 and 1
export function getNoise(point, scale, dimensions = point.length, evalAt = 0) {
    // note that the number of dimensions is decremented before it is used as an array index
    dimensions--;

    // the evalAt point represents one of the nearby corners to the point
    // it is compressed like a hash code into a single value
    evalAt = evalAt * scale[dimensions] + Math.floor(point[dimensions]);

    // if this is the last dimension, evaluate between two adjacent corners
    if (dimensions == 0) {
        return lerp(random(evalAt), random(evalAt + 1), point[0] % 1);
    }

    // otherwise recursively evaluate between two faces
    return lerp(
        getNoise(point, scale, dimensions, evalAt),
        getNoise(point, scale, dimensions, evalAt + 1),
        point[dimensions] % 1
    );
}

// output is consistent for values of x but varies greatly between nearby values
// returns a value between 0 and 1
function random(x) {
    // some bit twisting
    x = (x ^ 2747636419);
    x = (x * 2654435769);
    x = (x ^ (x >> 16));
    x = (x * 2654435769);
    x = (x ^ (x >> 16));
    x = (x * 2654435769);

    // get back to a 32-bit range
    x = (x % 0x100000000);

    // scale between 0 and 1
    return x / 0xffffffff;
}

// simple linear interpolation
function lerp(a, b, t) {
    return a + t * (b - a);
}
